Add StopOverseer() to close signal channel

kentlallen opened the pull request:
The signal channel is never closed so the goroutine to handle signals never exits.
This change adds a mechanism to close the channel and end the signal goroutine. 
This is useful for programs that use overseer for a single batch of processes and then don't use that overseer object again. Currently, each batch results in a zombie overseer signal handler process.

I won't mind if this isn't merged. I just thought the functionality might be useful to others and would be worth proposing.

croqaz commented:
Thank you for the PR @kentlallen ! The issue sounds pretty important to me, the manager should be completely closed when all processes are finished, or after StopAll.

I like the approach.
nitpick: if you can rename the sigChannel to signalChan (to match the other struct names: statusChan, changeChan and doneChan).
Also, if you can add a test to make sure it actually stops all processes when calling the new StopOverseer() function.

kentlallen commented:
Thanks for the feedback!

This update includes:
- rename sigChannel to signalChan
- Add tests for StopOverseer (includes adding defer StopOverseer to other tests to measure overseer procs accurately)
- Add auto CPU usage test for Linux. This test monitors the CPU usage for the specific process and _should_ be fairly hardware-independent. I added an environment variable that can be passed to the test to fine-tune this if needed.
- Bugfix to only start log processors when channels are not nil.

In writing these tests I found that when no output was selected the log goroutines would block indefinitely when ranging over the channel since the channel was nil. There are now checks for that case. I also added StopOverseer tests for stream, buffer, and no output to cover as many cases as possible.

I'm curious what the process is for releasing a new version. Is it as simple as updating the changelog and committing with a new version tag?

kentlallen commented:
> the manager should be completely closed when all processes are finished, or after StopAll.

As I've been thinking about this another option is to start the signal handler in `Supervise()/SuperviseAll()` instead of `NewOverseer()`. Then the channel could be closed at the end of `Supervise()`. That way it would be closed automatically instead of needing a new API call to do so. Also `NewOverseer()` wouldn't need to be called to start the signal handler again. Let me know what you think and I'll do some testing with that variation.

croqaz commented:
Hi! I am in a vacation currently and I can't clone the repo, or check the code. I'll check in a week or so, when I'm back. Thanks for the fixes and tests! Feel free to switch to the approach you think works best!

Mar 1, 2024 7:49:24 PM Kent Allen 'notifications at github.com[https://github.com]' ***@***.***>:

> 
> the manager should be completely closed when all processes are finished, or after StopAll.
> 
> As I've been thinking about this another option is to start the signal handler in *Supervise()/SuperviseAll()* instead of *NewOverseer()*. Then the channel could be closed at the end of *Supervise()*. That way it would be closed automatically instead of needing a new API call to do so. Also *NewOverseer()* wouldn't need to be called to start the signal handler again. Let me know what you think and I'll do some testing with that variation.
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/ShinyTrinkets/overseer/pull/26#issuecomment-1973626270], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AANK2XPE3TEJEGDVBDPZSSDYWC5RRAVCNFSM6AAAAABDV6RK2CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZTGYZDMMRXGA].
> You are receiving this because you commented.
> [Tracking image][https://github.com/notifications/beacon/AANK2XKNF73K6ACNORF35W3YWC5RRA5CNFSM6AAAAABDV6RK2CWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTVUMSZ4.gif]
>

kentlallen commented:
I hope you are enjoying your vacation!
I haven't been able to think of a good alternative way to start/stop the signal handler that works for `Supervise` as well as `SuperviseAll`.  I'll give it more thought, but the current implementation and this PR are the best I can come up with for the time being. 
Thanks for taking a look when you get a chance!

croqaz commented on manager.go:
```diff
@@ -95,12 +96,13 @@ func NewOverseer() *Overseer {
+		// Call StopOverseer() to exit this goroutine
+		for sig := range ovr.signalChan {
 			log.Error("Received signal: %v!", Attrs{"sig": sig})
 			ovr.StopAll(false)
```
In current implementation, calling StopAll(false) here is not needed, because StopOverseer() is required.
Maybe this logic here can be removed completely if we end up using a StopOverseer() function, this "go func" doesn't do anything special.

croqaz replied:
OK my mistake.... this logic here auto-stops the procs when receiving SIGINT or SIGTERM. It is useful, but if we use `StopOverseer()`, this is manual mode so we don't need to listen on signals anymore, this can be done outside of Overseer at the app level.

croqaz commented:
Hi @kentlallen ! Sorry for the late review. I checked the code and ran the tests. Super impressive work especially with the test that shows the CPU usage, I could see 1% usage during the test.

I'm not super convinced about the need to have a completely new *mandatory* function to shut down the Overseer... I liked the auto-shutdown method, even if it was a bit magical.

Also it looks like the `ovr.signalChan` channel is still used in the same places in the code like before and it is still used in the same way, to stop all child procs on shutdown, but now that we have this new *mandatory* function, is it needed anymore? I mean it still catches SIGINT and SIGTERM, but if the shutdown is manual, that's really needed.

About the release, I can do it after we merge this PR, it's just the changelog and a tag and a manual release from the tag.

croqaz commented on manager.go:
```diff
@@ -559,6 +565,19 @@ func (ovr *Overseer) StopAll(kill bool) {
 	ovr.access.Unlock()
```
One like above: `ovr.setState(IDLE)`;
If we end up using `StopOverseer()` function, we can't switch the Overseer back into idle state, because we want it to be one-use only.

kentlallen commented:
> I'm not super convinced about the need to have a completely new _mandatory_ function to shut down the Overseer... I liked the auto-shutdown method, even if it was a bit magical.
> 
> Also it looks like the `ovr.signalChan` channel is still used in the same places in the code like before and it is still used in the same way, to stop all child procs on shutdown, but now that we have this new _mandatory_ function, is it needed anymore? I mean it still catches SIGINT and SIGTERM, but if the shutdown is manual, it is duplicated functionality. Listening to signals can be done outside of Overseer, at the app level.

These are good points. I like the idea of closing everything automatically.

I think a good compromise might be to have the signal handler start when `SuperviseAll()` is called. Then, it can stop once `SuperviseAll()` has finished. This would provide the utility of the signal handler without having a _mandatory_ stop function. It would also allow an Overseer instance to be used multiple times since the handler would start with each call to `SuperiseAll()` instead of `NewOverseer()`.
The downside would be that users would have to handle SIGINT and SIGTERM at their app level when solely using `Supervise()`.

Thoughts?

kentlallen commented:
No problem about the back and forth! I think we've landed on a much better solution than I originally proposed. 

I have moved the signal handler into `SuperviseAll` and `signalChan` is once again local. Stopping the signal handler is automatic with a defer and `SuperviseAll` can once again be called multiple times. I also made the signal handler log the event via `ovr.loggers` so the watch logs will reflect that a SIGINT or SIGTERM has happened.

In addition, I added two new tests that exercise the signal handler directly by sending a SIGINT or SIGTERM to the overseer process.

Please let me know if there is anything else I can add or change and thank you again for taking the time to review my implementations!

croqaz commented:
Thank you very much for all the changes! I agree that the result looks much better after the back and forth! I love the implementation.
I will make a release soon.

croqaz commented:
The `getOverseerProcs` function only seems to work on Linux, but the CI also runs MacOS and those tests always fail, so I disabled them.
I think it's an acceptable compromise.

kentlallen commented:
Ah, that was an oversight. An alternative to simply disabling them all the time that I used for the auto CPU usage test:
`if runtime.GOOS != "linux" {
	t.Skip("Test only supported on Linux.")
}`

croqaz commented:
Yes, that's what I did! :) It's OK, it's not a bug.
